### PR TITLE
Improved performance of decoders

### PR DESCRIPTION
**Pull request by AlexKoff88**
Improved performance of decoders: GPT-like, bloom, etc., and Seq2seq models.
Observed a significant speedup of long sequences, e.g. Dolly-3B, 500 tokens: +30% when running on a client CPU.

### CODE REVIEW

File: optimum/intel/openvino/modeling_decoder.py [hunk 1]
```diff
@@ -301,15 +308,14 @@ def forward(
             inputs["attention_mask"] = np.array(attention_mask)
 
         # Run inference
-        outputs = self.request(inputs, shared_memory=True)
+        self.request.start_async(inputs, shared_memory=True)
```
Comment (helena-intel) [inline optimum/intel/openvino/modeling_decoder.py:311]: Is there any danger in using shared_memory in combination with async in this case?

Comment (AlexKoff88) [reply optimum/intel/openvino/modeling_decoder.py:311]: it is not truly the async since we wait for the results right after the call, it is more a way to access the results without copying them (this is what you get using the current synchronous API).

File: optimum/intel/openvino/quantization.py [hunk 2]
```diff
@@ -234,17 +235,26 @@ def __init__(self, request):
...
             def __getattr__(self, attr):
                 if attr in self.__dict__:
                     return getattr(self, attr)
                 return getattr(self.request, attr)
 
         self.model.request = InferRequestWrapper(self.model.request)
-        for i, data in enumerate(calibration_dataloader):
-            self.model.generate(**data, max_new_tokens=10)
+        for _, data in enumerate(calibration_dataloader):
+            self.model.generate(**data, max_new_tokens=100)
```
Comment (echarlaix) [inline optimum/intel/openvino/quantization.py:257]: Is this modification added to reduce accuracy degradation resulting from quantization? If yes, what did you observe when varying this parameter ?

File: optimum/intel/openvino/quantization.py [hunk 3]
```diff
@@ -234,17 +235,26 @@ def __init__(self, request):
...
+
             def __getattr__(self, attr):
                 if attr in self.__dict__:
                     return getattr(self, attr)
                 return getattr(self.request, attr)
 
         self.model.request = InferRequestWrapper(self.model.request)
-        for i, data in enumerate(calibration_dataloader):
-            self.model.generate(**data, max_new_tokens=10)
+        for _, data in enumerate(calibration_dataloader):
```
Comment (echarlaix) [inline optimum/intel/openvino/quantization.py:256]: Not related to the PR, but what do you think about uniformizing how quantization is applied on causal langage models depending on whether the user gives a `torch.nn.Module` or a `OVBaseDecoderModel` (the number of generation steps is currently not the same). We could also instantiate an `OVModel` in the `from_pretrained` method when the given model is a `PreTrainedModel`

**Review by echarlaix — APPROVED** [review]

File: optimum/intel/openvino/modeling_decoder.py [hunk 4]
```diff
@@ -237,6 +237,13 @@ def reshape(self, batch_size: int, sequence_length: int):
         logger.warning("Static shapes are not supported for causal language model.")
         return self
 
+    def compile(self):
+        if self.request is None:
+            logger.info("Compiling the model...")
+            cache_dir = Path(self.model_save_dir).joinpath("model_cache")
+            ov_config = {**self.ov_config, "CACHE_DIR": str(cache_dir)}
+            self.request = core.compile_model(self.model, self._device, ov_config).create_infer_request()
```
Comment (echarlaix) [inline optimum/intel/openvino/modeling_decoder.py:245]: could be replaced by : 
```suggestion
            super().compile()
            self.request = self.request.create_infer_request()
```

File: optimum/intel/openvino/quantization.py (hunk 2 above)
Comment (AlexKoff88) [reply optimum/intel/openvino/quantization.py:257]: In the process, and will update a bit later.

File: optimum/intel/openvino/modeling_decoder.py (hunk 1 above)
Comment (helena-intel) [reply optimum/intel/openvino/modeling_decoder.py:311]: Ah, thanks! This is a great solution for now. I heard sync inference in OpenVINO should be improved in the next release, with an option to not return the data.

File: optimum/intel/openvino/quantization.py (hunk 3 above)
Comment (AlexKoff88) [reply optimum/intel/openvino/quantization.py:265]: It is hard to accomplish this with the current NNCF PTQ API implementation we have for PyTorch. I think we should deprecate PTQ for PyTorch at some point because it also introduces ambiguity for the user about what workflow to use for quantization.

**Comment by AlexKoff88** [comment]
I think it is ready for merge.

**Review by helena-intel — APPROVED** [review]
🚀 🔥

**Review by echarlaix — APPROVED** [review]
LGTM, thanks a lot @AlexKoff88
